hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
-const auto SLOW_ADVERT_INTERVAL = MSEC_TO_UNITS(500, UNIT_0_625_MS); // Slow mode
+system_error_t sysError(uint32_t error) {
+    switch (error) {
+        case NRF_SUCCESS:                       return SYSTEM_ERROR_NONE;
+        case NRF_ERROR_SVC_HANDLER_MISSING:
+        case NRF_ERROR_NOT_SUPPORTED:           return SYSTEM_ERROR_NOT_SUPPORTED;
+        case NRF_ERROR_SOFTDEVICE_NOT_ENABLED:
+        case NRF_ERROR_INVALID_STATE:           return SYSTEM_ERROR_INVALID_STATE;
+        case NRF_ERROR_INTERNAL:                return SYSTEM_ERROR_INTERNAL;
+        case NRF_ERROR_NO_MEM:                  return SYSTEM_ERROR_NO_MEMORY;
+        case NRF_ERROR_NOT_FOUND:               return SYSTEM_ERROR_NOT_FOUND;
+        case NRF_ERROR_INVALID_PARAM:
+        case NRF_ERROR_INVALID_LENGTH:
+        case NRF_ERROR_INVALID_FLAGS:
+        case NRF_ERROR_INVALID_DATA:
+        case NRF_ERROR_DATA_SIZE:
+        case NRF_ERROR_NULL:                    return SYSTEM_ERROR_INVALID_ARGUMENT;
+        case NRF_ERROR_TIMEOUT:                 return SYSTEM_ERROR_TIMEOUT;
+        case NRF_ERROR_FORBIDDEN:
+        case NRF_ERROR_INVALID_ADDR:            return SYSTEM_ERROR_NOT_ALLOWED;
+        case NRF_ERROR_BUSY:                    return SYSTEM_ERROR_BUSY;
+        case NRF_ERROR_CONN_COUNT:              return SYSTEM_ERROR_LIMIT_EXCEEDED;
+        case NRF_ERROR_RESOURCES:               return SYSTEM_ERROR_ABORTED;
+        default:                                return SYSTEM_ERROR_UNKNOWN;
+    }
+}
+
+bool addressEqual(const hal_ble_addr_t& srcAddr, const hal_ble_addr_t& destAddr) {
+    return (srcAddr.addr_type == destAddr.addr_type && !memcmp(srcAddr.addr, destAddr.addr, BLE_SIG_ADDR_LEN));
+}
+
+} //anonymous namespace
+
+class GattBase {
+public:
+    GattBase() {
+        desiredAttMtu_ = BLE_MAX_ATT_MTU_SIZE;
+        attMtuList_.clear();
+    }
+    ~GattBase() = default;
+
+protected:
+    /*
+     * A device's Exchange MTU Request shall contain the same MTU as the
+     * device's Exchange MTU Response (i.e. the MTU shall be symmetric).
+     */
+    class LinkAttMtu {
+    public:
+        uint16_t connHandle;
+        uint16_t effective;
+        bool isExchanged;
+        LinkAttMtu() {
+            connHandle = BLE_INVALID_CONN_HANDLE;
+            effective = BLE_DEFAULT_ATT_MTU_SIZE;
+            isExchanged = false;
+        }
+        ~LinkAttMtu() = default;
+    };
+
+    // GATT Server and GATT client share the same ATT_MTU.
+    static uint16_t desiredAttMtu_;
+    static Vector<LinkAttMtu> attMtuList_;
+
+    size_t getAttMtu(uint16_t connHandle) {
+        for (int i = 0; i < attMtuList_.size(); i++) {
+            LinkAttMtu& attMtu = attMtuList_[i];
+            if (attMtu.connHandle == connHandle) {
+                return attMtu.effective;
+            }
+        }
+        return 0;
+    }
+
+    void configureAttMtu(uint16_t connHandle, uint16_t effective, bool exchanged=false) {
```

Comment:
Nitpicking: prefer `bool exchanged = false` formatting for default argument values.